### PR TITLE
Run kind e2e tests

### PR DESCRIPTION
**Pull request by tcnghia**
# Description
This workflow runs our e2e tests in a local KinD cluster. Since it does not required a paid cluster and special credentials, it does not require /ok-to-test. That allows contributors to run E2E tests in their Draft PRs without explicit approval from Dapr
maintainers.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: N/A

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: .github/workflows/kind-e2e.yaml [hunk 1]
```diff
@@ -0,0 +1,156 @@
+# ------------------------------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------------------------------
+name: KinD tests
```
Comment (youngbupark) [inline .github/workflows/kind-e2e.yaml:5]: Can we make this workflow name explicitly? e.g. `E2E test on KinD`

Comment (tcnghia) [reply .github/workflows/kind-e2e.yaml]: Done

File: .github/workflows/kind-e2e.yaml [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            endpoint = ["http://$REGISTRY_NAME:$REGISTRY_PORT"]
+        EOF
+
+        # Log the generated kind.yaml for easy reference.
+        cat kind.yaml
+      
+    - name: Create KinD Cluster
+      uses: helm/kind-action@v1.0.0
+      with:
+        config: kind.yaml
```
Comment (youngbupark) [inline .github/workflows/kind-e2e.yaml:103]: Please specify the exact version of kind here.

File: .github/workflows/kind-e2e.yaml [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    - name: Setup local registry
+      run: |
+        # Run a registry.
+        docker run -d --restart=always \
+          -p $REGISTRY_PORT:$REGISTRY_PORT --name $REGISTRY_NAME registry:2
+        # Connect the registry to the KinD network.
+        docker network connect "kind" $REGISTRY_NAME
+        
+    - name: Setup Helm
+      uses: azure/setup-helm@v1
```
Comment (youngbupark) [inline .github/workflows/kind-e2e.yaml:123]: please specify helm version.

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: .github/workflows/kind-e2e.yaml (hunk 2 above)
Comment (tcnghia) [reply .github/workflows/kind-e2e.yaml:103]: Done

File: .github/workflows/kind-e2e.yaml (hunk 3 above)
Comment (tcnghia) [reply .github/workflows/kind-e2e.yaml:124]: Done

**Review by youngbupark — DISMISSED** [review]

**Review by artursouza — DISMISSED** [review]

**Comment by youngbupark** [comment]
this pr https://github.com/dapr/dapr/pull/2148 is merged. so it will make test a bit faster.

File: tests/dapr_tests.mk [hunk 4]
```diff
@@ -119,7 +119,7 @@ test-e2e-all: check-e2e-env
 	# tests. In the future, if we add any tests that modify global state (such as dapr config), we'll 
 	# have to be sure and run them after the main test suite, so as not to alter the state of a running
 	# test
-	GOOS=$(TARGET_OS_LOCAL) DAPR_TEST_NAMESPACE=$(DAPR_TEST_NAMESPACE) DAPR_TEST_TAG=$(DAPR_TEST_TAG) DAPR_TEST_REGISTRY=$(DAPR_TEST_REGISTRY) DAPR_TEST_MINIKUBE_IP=$(MINIKUBE_NODE_IP) go test -p 2 -count=1 -v -tags=e2e ./tests/e2e/...
+	GOOS=$(TARGET_OS_LOCAL) DAPR_TEST_NAMESPACE=$(DAPR_TEST_NAMESPACE) DAPR_TEST_TAG=$(DAPR_TEST_TAG) DAPR_TEST_REGISTRY=$(DAPR_TEST_REGISTRY) DAPR_TEST_MINIKUBE_IP=$(MINIKUBE_NODE_IP) go test -p 2 -count=1 -v -tags=e2e ./tests/e2e/... -timeout 20m
```
Comment (mukundansundar) [inline tests/dapr_tests.mk:122]: will it timeout for `ok-to-test` in also 20 minutes?

Comment (youngbupark) [reply tests/dapr_tests.mk:122]: oh nice catch.. we should remove the timeout.

Comment (tcnghia) [reply tests/dapr_tests.mk]: Removed the timeout. Thanks

File: tests/e2e/utils/helpers.go [hunk 5]
```diff
@@ -136,6 +172,16 @@ func HTTPGetRawNTimes(url string, n int) (*http.Response, error) {
 	return res, err
 }
 
+// httpGetRaw is a helper to make GET request call to url
+func httpGetRaw(client http.Client, url string) (*http.Response, error) {
```
Comment (mukundansundar) [inline tests/e2e/utils/helpers.go:176]: modify function `HTTPGetRaw` to use this function?

**Comment by youngbupark** [comment]
@tcnghia please fix linter errors.

File: tests/e2e/utils/helpers.go (hunk 5 above)
Comment (tcnghia) [reply tests/e2e/utils/helpers.go:177]: Done

**Comment by tcnghia** [comment]
@artursouza @youngbupark @mukundansundar I tweaked a few more things to make the workflow a bit more stable than a few days ago. Please take another look. I would love to give it a try for more data about reliability/stability.

We can make it a required workflow if/when it is solid.

**Review by youngbupark — DISMISSED** [review]

File: tests/e2e/runtime/runtime_test.go [hunk 6]
```diff
@@ -49,16 +47,9 @@ func getAPIResponse(t *testing.T, testName, runtimeExternalURL string) (*daprAPI
 	// this is the publish app's endpoint, not a dapr endpoint
 	url := fmt.Sprintf("http://%s/tests/%s", runtimeExternalURL, testName)
 
-	resp, err := http.Get(url)
-	defer resp.Body.Close()
-
+	body, err := utils.HTTPGet(url)
 	require.NoError(t, err)
-	require.Equal(t, resp.StatusCode, http.StatusOK)
```
Comment (mukundansundar) [inline tests/e2e/runtime/runtime_test.go:56]: do not remove the status code check.

Comment (tcnghia) [reply tests/e2e/runtime/runtime_test.go:56]: The return value of `HTTPGet` isn't http.Response any more but just byte[]

Comment (tcnghia) [reply tests/e2e/runtime/runtime_test.go:56]: Switch to `HTTPGetRaw` to keep the same logic. I only moved the `nil` deref down.

**Comment by tcnghia** [comment]
Windows workflow failing due to clusters being busy

```
Scanning dapr-aks-e2e-08 ...
Switching to dapr-aks-e2e-08 context...
Merged "dapr-aks-e2e-08" as current context in C:\Users\runneradmin\.kube\config
Trying to create dapr-tests namespace...
Error from server (AlreadyExists): namespaces "dapr-tests" already exists
Namespace is being used for 58 seconds
-------------------------------------------------------
All test clusters are fully occupied.
```

Linux workflow was canceled because of that.

Trying again...

/ok-to-test

File: tests/config/kafka_override.yaml [hunk 7]
```diff
@@ -25,6 +25,7 @@ affinity:
             values:
             - linux
 zookeeper:
+  replicaCount: 1
```
Comment (mukundansundar) [inline tests/config/kafka_override.yaml:28]: @yaron2  @youngbupark @wcs1only are we ok with reducing the replicaCount to 1?

Comment (tcnghia) [reply tests/config/kafka_override.yaml:28]: I'm fine with keeping the existing count too, it's not required for this PR - just improving the workflow time a little bit.

Comment (youngbupark) [reply tests/config/kafka_override.yaml:28]: this must be ok.

**Review by youngbupark — APPROVED** [review]

**Pull request merged**